From fe1fe76424fe619267fbbaf214c67262b9c4997c Mon Sep 17 00:00:00 2001
From: Marek Marczykowski <marmarek@invisiblethingslab.com>
Date: Wed, 3 Apr 2013 17:08:54 +0200
Subject: [PATCH 01/17] libxl: allow only 'ethernet' and 'bridge' interfaces,
 allow script there
Organization: Invisible Things Lab
Cc: Marek Marczykowski <marmarek@invisiblethingslab.com>

Actually only those interface types are handled correctly so reject
others instead of ignoring settings (i.e. treating as bridge/ethernet
anyway).
Also allow <script/> in 'ethernet' (which should be the only
script-allowing type). Keep <script/> allowed in bridge to be compatible
with legacy 'xen' driver.

Changes in v2:
 - reject interfaces other than 'ethernet' or 'bridge'
 - change title to better match patch content (was "libxl: allow script
   for any network interface, not only bridge")
 - update description

Signed-off-by: Marek Marczykowski <marmarek@invisiblethingslab.com>
---
 src/libxl/libxl_conf.c | 36 +++++++++++++++++++-----------------
 1 file changed, 19 insertions(+), 17 deletions(-)

diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index 3de642b..648d399 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -589,24 +589,26 @@ libxlMakeNic(virDomainNetDefPtr l_nic, libxl_device_nic *x_nic)
         return -1;
     }
 
-    if (l_nic->type == VIR_DOMAIN_NET_TYPE_BRIDGE) {
-        if (l_nic->data.bridge.brname &&
-            (x_nic->bridge = strdup(l_nic->data.bridge.brname)) == NULL) {
-            virReportOOMError();
-            return -1;
-        }
-        if (l_nic->script &&
-            (x_nic->script = strdup(l_nic->script)) == NULL) {
-            virReportOOMError();
-            return -1;
-        }
-    } else {
-        if (l_nic->script) {
-            virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
-                           _("scripts are not supported on interfaces of type %s"),
-                           virDomainNetTypeToString(l_nic->type));
+    switch (l_nic->type) {
+        case VIR_DOMAIN_NET_TYPE_BRIDGE:
+            if (l_nic->data.bridge.brname &&
+                    (x_nic->bridge = strdup(l_nic->data.bridge.brname)) == NULL) {
+                virReportOOMError();
+                return -1;
+            }
+            /* fallthrough */
+        case VIR_DOMAIN_NET_TYPE_ETHERNET:
+            if (l_nic->script &&
+                    (x_nic->script = strdup(l_nic->script)) == NULL) {
+                virReportOOMError();
+                return -1;
+            }
+            break;
+        default:
+            virReportError(VIR_ERR_INTERNAL_ERROR,
+                    _("libxenlight does not support network device type %s"),
+                    virDomainNetTypeToString(l_nic->type));
             return -1;
-        }
     }
 
     return 0;
-- 
1.8.1.4

